File: Problems/python/boj_정수삼각형.py
#https://www.acmicpc.net/problem/1932
#DP

import sys
input = sys.stdin.readline

n = int(input())
l = []
for _ in range(n):
    l.append(list(map(int, input().split())))

for i in range(1, n):
    for j in range(len(l[i])):
        if j==0:
            l[i][j] += l[i-1][0]
        elif j==len(l[i])-1:
            l[i][j] += l[i-1][-1]
        else:
            l[i][j] += max(l[i-1][j-1], l[i-1][j])

print(max(l[-1]))

'''
DP table 변수를 따로 만드려하니 어떻게 만들어야할지 막막했다.
그리고 그것에 맞춰 문제를 풀려다보니 잘못된 접근이 계속 반복됐던 것 같다.

현재 입력을 받은 배열을 그대로 사용하여 풀었다.
그로 인해 행 번호가 증가하며 진행되는 상황을 누적합을 이용하여 표현할 수 있었다.
'''

File: Problems/cpp/boj_정수삼각형.cpp
//https://www.acmicpc.net/problem/1932
//https://school.programmers.co.kr/learn/courses/30/lessons/43105
//DP

//거꾸로부터 올라가는 것. 정답에 최댓값 안찾아도 돼서 편함.
#include <vector>
#include <algorithm>

using namespace std;

int solution(vector<vector<int>> tri) {
    int answer;
    for (int i=tri.size()-1; i>0; i--){
        for (int j=0; j<tri[i].size()-1; j++){
            if (tri[i][j] > tri[i][j+1])
                tri[i-1][j] += tri[i][j];
            else
                tri[i-1][j] += tri[i][j+1];
        }
    }
    
    answer = tri[0][0];
    
    return answer;
}

//첫 번째 풀이
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;
 
int main(){
    ios::sync_with_stdio(0);
    cin.tie(0); cout.tie(0);

    int n;
    cin >> n;

    vector<vector<int>> tri;
    for (int i=0; i<n; i++){
        vector<int> tmp(i+1);
        for (int j=0; j<i+1; j++)
            cin >> tmp[j];

        tri.push_back(tmp);
    }

    for (int i=1; i<n; i++){
        for (int j=0; j<tri[i].size(); j++){
            if (j==0)
                tri[i][j] += tri[i-1][j];
            else if (j==tri[i].size()-1)
                tri[i][j] += tri[i-1][j-1];
            else
                tri[i][j] += max(tri[i-1][j-1], tri[i-1][j]);
        }
    }

    cout << *max_element(tri[n-1].begin(), tri[n-1].end());
}